src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -340,221 +228,130 @@ void OnBrowseAdd(BrowseContext * context, const char * name, const char * type,
-    service.mTextEntrySize = sdCtx->textEntries.empty() ? 0 : sdCtx->textEntries.size();
-    // Use the first IP we got for the DnssdService.
-    if (sdCtx->addresses.size() != 0)
+    if (!(flags & kDNSServiceFlagsMoreComing))
     {
-        service.mAddress.SetValue(sdCtx->addresses.front());
-        sdCtx->addresses.erase(sdCtx->addresses.begin());
+        VerifyOrReturn(sdCtx->HasAddress(), sdCtx->Finalize(kDNSServiceErr_BadState));
+        sdCtx->Finalize();
     }
-    Platform::CopyString(service.mName, sdCtx->name);
-    Platform::CopyString(service.mHostName, hostname);
-    service.mInterface = Inet::InterfaceId(sdCtx->interfaceId);
-
-    // TODO: Does it really make sense to pass in the status from our last "get the IP address" operation?
-    sdCtx->callback(sdCtx->context, &service, Span<Inet::IPAddress>(sdCtx->addresses.data(), sdCtx->addresses.size()), status);
-    MdnsContexts::GetInstance().Remove(sdCtx);
 }
 
-static CHIP_ERROR GetAddrInfo(void * context, DnssdResolveCallback callback, uint32_t interfaceId,
-                              chip::Inet::IPAddressType addressType, const char * name, const char * hostname, uint16_t port,
-                              uint16_t txtLen, const unsigned char * txtRecord)
+static void GetAddrInfo(ResolveContext * sdCtx)
 {
-    DNSServiceErrorType err;
-    DNSServiceRef sdRef;
-    GetAddrInfoContext * sdCtx;
-
-    sdCtx = chip::Platform::New<GetAddrInfoContext>(context, callback, name, interfaceId, port);
+    auto protocol = sdCtx->protocol;
 
-    char key[kDnssdKeyMaxSize];
-    char value[kDnssdTextMaxSize];
-    uint8_t valueLen;
-    const void * valuePtr;
-
-    uint16_t recordCount = TXTRecordGetCount(txtLen, txtRecord);
-    for (uint16_t i = 0; i < recordCount; i++)
+    for (auto & service : sdCtx->services)
     {
-        err = TXTRecordGetItemAtIndex(txtLen, txtRecord, i, kDnssdKeyMaxSize, key, &valueLen, &valuePtr);
-        VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
-
-        if (valueLen >= sizeof(value))
-        {
-            // Truncation, but nothing better we can do
-            valueLen = sizeof(value) - 1;
-        }
-        memcpy(value, valuePtr, valueLen);
-        value[valueLen] = 0;
-
-        sdCtx->textEntries.push_back(TextEntry{ strdup(key), reinterpret_cast<const uint8_t *>(strdup(value)), valueLen });
-    }
-
-    DNSServiceProtocol protocol;
-
-#if INET_CONFIG_ENABLE_IPV4
-    if (addressType == chip::Inet::IPAddressType::kIPv4)
-    {
-        protocol = kDNSServiceProtocol_IPv4;
-    }
-    else if (addressType == chip::Inet::IPAddressType::kIPv6)
-    {
-        protocol = kDNSServiceProtocol_IPv6;
-    }
-    else
-    {
-        protocol = kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6;
+        auto interfaceId = service.first;
+        auto hostname    = GetHostNameWithDomain(service.second.mHostName);
```

Comment:
Unit tests for dnssd are timing out because `service.second.mHostName` is not long enough to hold the hostname we try to use, and when we copy the hostname into it we just silently truncate.

We should consider erroring out loudly at that point, logging the actual input into `DNSServiceGetAddrInfo` we use here, and figure out how to either fix the hostname we use or stop using the DnssdService to store it for now...

Reply:
I have introduce a `InterfaceInfo` class that holds the hostname, the service and the resolved ids.